plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php

```diff
@@ -0,0 +1,300 @@
+<?php
+/**
+ * Controller to manage processing of `BatchProcessor` type. Trigger the `enqueue_processor` method to start a batch process, and it will manage rest of the things.
+ */
+
+namespace Automattic\WooCommerce\Internal\BatchProcessing;
+
+use Dompdf\Exception;
+
+/**
+ * Class BatchProcessingController
+ *
+ * @package Automattic\WooCommerce\Internal\Updates.
+ */
+class BatchProcessingController {
+
+	const PENDING_PROCESSES_OPTION_NAME = 'wc_pending_batch_processes';
+	const CONTROLLER_CRON_NAME          = 'wc_schedule_pending_batch_processes';
+	const SINGLE_BATCH_PROCESS_ACTION   = 'wc_run_batch_process';
+	const ACTION_GROUP                  = 'wc_batch_processes';
+
+	/**
+	 * BatchProcessingController constructor.
+	 *
+	 * Schedules necessary actions to process batches.
+	 */
+	public function __construct() {
+		add_action(
+			self::CONTROLLER_CRON_NAME,
+			function () {
+				$this->schedule_processes();
+			}
+		);
+		add_action(
+			self::SINGLE_BATCH_PROCESS_ACTION,
+			function ( $batch_process ) {
+				$this->process_single_batch( $batch_process );
+			},
+			10,
+			2
+		);
+	}
+
+	/**
+	 * Starts an update.
+	 *
+	 * @param string $processor_class_name Fully qualified class name of the processor, must be child class of `BatchProcessor`.
+	 */
+	public function enqueue_processor( string $processor_class_name ) {
```

Comment:
I have a question about this. Imagine this scenario:

1. `enqueue_processor` is invoked, this schedules `CONTROLLER_CRON_NAME` to be run immediately
2. `schedule_processes` runs, this schedules `CONTROLLER_CRON_NAME` again but for one hour later
3. Before the hour passes, `enqueue_processor` is invoked, scheduling `CONTROLLER_CRON_NAME`again

What happens then? The last scheduled action overrides the first one, or both keep scheduled and run at their proper times? If so, shouldn't we be checking if the action is already scheduled before scheduling a new one, to prevent a neverending stream of "action lanes" (one per execution of `enqueue_processor`) that would keep rescheduling themselves over and over?

Reply:
You are right, seems like I had missed this. I have added a `unique` param in https://github.com/woocommerce/woocommerce/pull/33233/commits/4be2dd5e3d88ed3883deb5aa41ac2e1a2ec98982 when we first start the process from `enqueue_processor` (but not on the re-schedule ones since they are still running and will be excluded). Note that this is a fairly new API that has not been released yet, although it's merged, so if you want to test you would have checkout the latest trunk in Action Scheduler.